deepspeed/comm/ccl.py

```diff
@@ -39,27 +46,100 @@ def __init__(self, name='ccl', rank=-1, world_size=-1, mpu=None, timeout=None, i
         size = self.get_world_size()
         rank = self.get_rank()
         main_kvs = self.ccl_comm_op.get_kvs_addr(rank)
-        main_kvs = torch.tensor(main_kvs).to(torch.uint8)
+        main_kvs = torch.tensor(main_kvs).to(torch.uint8).to(get_accelerator().device_name(rank))
         super(CCLBackend, self).broadcast(main_kvs, 0)
         self.ccl_comm_op.initialize(size, rank, main_kvs)
         self.initialized = True
+        self.available_coll = self.ccl_comm_op.get_available_coll()
 
     def is_initialized(self):
         return self.initialized
-
-    def broadcast(self, tensor, src, group=None, async_op=False):
-        self.ccl_comm_op.broadcast(tensor, src, group, async_op)
-
+    
+    def run_collective(self, name, **kwargs):
+        if name in self.available_coll:
+            kwargs['group'] = self.get_all_ranks_from_group(kwargs['group'])
+            func = "self.ccl_comm_op." + name
+            eval(func)(*(kwargs.values()))
+            return CCLHandler(self.ccl_comm_op)
+        else:
+            func = "super(CCLBackend, self)." + name
+            return eval(func)(*(kwargs.values()))
+        
     def all_reduce(self, tensor, op=ReduceOp.SUM, group=None, async_op=False):
         use_caching = False
         if use_caching:
             match_id = f"{tensor.size()}-{op}"
-            self.ccl_comm_op.all_reduce_caching(tensor, op, match_id, group, async_op)
+            return self.run_collective(name="all_reduce_caching", tensor=tensor, op=op, match_id=match_id, group=group, async_op=async_op)
         else:
-            self.ccl_comm_op.all_reduce(tensor, op, group, async_op)
+            return self.run_collective(name="all_reduce", tensor=tensor, op=op, group=group, async_op=async_op)
 
     def inference_all_reduce(self, tensor, op=ReduceOp.SUM, group=None, async_op=False):
-        self.ccl_comm_op.inference_all_reduce(tensor, op, group, async_op)
+        return self.run_collective(name="inference_all_reduce", tensor=tensor, op=op, group=group, async_op=async_op)
+
+    def broadcast(self, tensor, src, group=None, async_op=False):
+        return self.run_collective(name="broadcast", tensor=tensor, src=src, group=group, async_op=async_op)     
+
+    def all_gather(self, tensor_list, tensor, group=None, async_op=False):    
+        return self.run_collective(name="all_gather", tensor_list=tensor_list, tensor=tensor, group=group, async_op=async_op)
+
+    def reduce_scatter_tensor(self, output_tensor, input_tensor, op, group=None, async_op=False):
+        return self.run_collective(name="reduce_scatter_tensor", output_tensor=output_tensor, input_tensor=input_tensor, op=op, group=group)
+
+    def all_gather_into_tensor(self, output_tensor, input_tensor, group=None, async_op=False):
```

Comment:
FYI @Liangliang-Ma - you'll want to run our formatter on this code to ensure it passes that check.

Reply:
@loadams Thanks! I corrected the format now.